### ISSUE TITLE
Modifying a configuration after load

### ISSUE THREAD

**Issue by muliyul** [swagger-api/swagger-core#3861]
Is there a hook for doing post config initialization?

My use case:
I'd like to provide users of my library the option to specify `readerClass` but if they don't provide one I'd like to use a fallback reader.

Is this possible with current api?

**Comment by frantuma** [comment]
Not sure about your scenario, and how are you using swagger-core / swagger-jaxrs; in "standard" configuration, if you don't provide a `readerClass` the standard `Reader` is used. This is true for any way you configure (via config file, problematically, etc.). If you need to customize this behaviour (if I get this right, if your user doesn't provide a `readerClass` you want to use your own one, not the `swagger-core` provided standard), you can use programmatic config or any other means to build dynamically a configuration to pass to Swagger, please see [wiki](https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-Configuration#configuration)

Please close ticket if this answers your question

**Comment by muliyul** [comment]
> Not sure about your scenario, and how are you using swagger-core / swagger-jaxrs; in "standard" configuration, if you don't provide a `readerClass` the standard `Reader` is used. This is true for any way you configure (via config file, problematically, etc.). If you need to customize this behaviour (if I get this right, if your user doesn't provide a `readerClass` you want to use your own one, not the `swagger-core` provided standard), you can use programmatic config or any other means to build dynamically a configuration to pass to Swagger, please see [wiki](https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-Configuration#configuration)
> 
> Please close ticket if this answers your question

Not quite. I'd like to preserve default scanning (of openapi.[json|yaml]) but override the reader class.

**Comment by frantuma** [comment]
Still I am not sure about your scenario, can you provide some more details? e.g. what do you mean `I'd like to preserve default scanning (of openapi.[json|yaml])`? and how are you configuring `swagger-core`, in which environment? some code samples demonstrating your env and goal would also be handy

**Comment by muliyul** [comment]
I'm using the JAX-RS programmatic implementation. In order to override the reader class I need to provide a configuration via `.openApiConfiguration(cfg)` but that way I lose the `openapi.yaml` (or `openapi.json`) reading support from the classpath and I need to read them manually (write code that reads them). I'd like to be able to get an instance of `SwaggerConfiguration` after it's being read (parsed) and modify it (that's what I meant by post config initialization).

Here's some code sample for reference:

```kotlin
JaxrsOpenApiContextBuilder()
                    .servletConfig(servletConfig)
                    .application(this)
                    .openApiConfiguration(swaggerConfig.apply { 
                     // I have to instantiate this swagger config somehow but I want to preserve the default behavior
                     // e.g reading from known locations etc...
                     readerClass = readerClass ?: "com.something.CustomReader"
                    })
                    .buildContext(true);
```

**Comment by frantuma** [comment]
you can provide the reader to the context according to your rules, and then init the context, the provided reader takes precedence over the one specified in configuration (or if empty, over the default one):

```
..context = new JaxrsOpenApiContextBuilder()
                    .servletConfig(servletConfig)
                    .application(this)
                    .buildContext(false);  // init FALSE

if (whatever your logic) {
	context.setOpenApiReader(myReaderInstance);
}
context.init();
```

**Comment by muliyul** [comment]
Awesome I'll try it and close if no more issues 👍

**Comment by muliyul** [comment]
This didn't work for me. Configuration is still null here.

```kotlin
JaxrsOpenApiContextBuilder<JaxrsOpenApiContextBuilder<*>>()
			.servletConfig(servletConfig)
			.application(environment.jersey().resourceConfig)
			.resourcePackages(resourcesPackageNames)
			.ctxId(ServletConfigContextUtils.getContextIdFromServletConfig(servletConfig))
			.buildContext(false)
			.apply { // this = JaxrsOpenApiContextBuilder
				if(this.openApiConfiguration?.readerClass.isNullOrEmpty()) {
					setOpenApiReader(DropwizardCompatibleReader()) // always executes because openApiConfiguration is always null
				}
			}
			.init()
```

**Comment by frantuma** [comment]
Right, possibly now use case is getting more clear, probably the easiest way to achieve that is something like:

```
..context = new JaxrsOpenApiContextBuilder()
                    .servletConfig(servletConfig)
                    .application(this)
                    .buildContext(true);  // init True, config loaded if present

if (context.getOpenAPIConfiguration() != null && StringUtils.isEmpty(context.getOpenAPIConfiguration().getReaderClass())) {
	context.setOpenApiReader(myReaderInstance);
}
...
context.read()
```

Alternatively slightly more complicated, as `loadConfiguration` method in `context` is protected, you could subclass the context and override `loadConfiguration`, then invoke that and you have the loaded configuration that you can query etc.

**Comment by muliyul** [comment]
> Right, possibly now use case is getting more clear, probably the easiest way to achieve that is something like:
> 
> ```
> ..context = new JaxrsOpenApiContextBuilder()
>                     .servletConfig(servletConfig)
>                     .application(this)
>                     .buildContext(true);  // init True, config loaded if present
> 
> if (context.getOpenAPIConfiguration() != null && StringUtils.isEmpty(context.getOpenAPIConfiguration().getReaderClass())) {
> 	context.setOpenApiReader(myReaderInstance);
> }
> ...
> context.read()
> ```
> 
> Alternatively slightly more complicated, as `loadConfiguration` method in `context` is protected, you could subclass the context and override `loadConfiguration`, then invoke that and you have the loaded configuration that you can query etc.

I don't understand you suggestion. can you give an example?

**Comment by frantuma** [comment]
An example of what? of the second alternative? the first one would match your use case I guess

**Comment by muliyul** [comment]
Yup the second one.

**Comment by muliyul** [comment]
Let me rephrase:

I have this code:
```kotlin
val openApi = JaxrsOpenApiContextBuilder<JaxrsOpenApiContextBuilder<*>>()
			.application(environment.jersey().resourceConfig)
			.resourcePackages(resourcesPackageNames)
			.buildContext(false)
			.apply {
				setOpenApiScanner(CompatibleScanner())
				setOpenApiReader(CompatibleReader())
			}
			.init()
			.read()
```

I expect `openApi` to contain a merging of both JAX-RS resource definitions and `openapi,yaml` definitions however this is not the case. How to achieve this? (I want to use the default lookups for the `yaml` definitions as described [here](https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-Configuration#known-locations)
